geodist: use Point struct instead of a Point interface

otan opened the pull request:
Getting an Edge is expensive as the generating a Point that complies
with the geodist.Point interface involves mallocing from memory. We call
Edge(x) a lot when doing POLYGON <=> POLYGON comparisons.

Instead, change the Point to a struct, which does not require
any malloc to generate interfaces at the cost of code ugliness.

Without bounding box checks, this speeds up a query from 2min to
1min30sec on a ~120000 row join.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50790)
<!-- Reviewable:end -->

sumeerbhola reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50790)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geodist/geodist.go, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50790#-MB5totP2bpWBWm-Tc7x:-MB5totP2bpWBWm-Tc7y:b-q4qb75) ([raw file](https://github.com/cockroachdb/cockroach/blob/89a3dfa2466e9cc05923c9b2e68ce9631f2de6b2/pkg/geo/geodist/geodist.go#L32)):*
> ```Go
> type PointUnion struct {
> 	GeomCoord geom.Coord
> 	GeogPoint s2.Point
> ```

This seems a bit messy, and we're paying the cost of a struct with 3 floats (the `s2.Point`) on the geometry path.

It seems to me that the reason for the allocation is that the implementations of `Point` use a pointer receiver. If we are going to be copying `PointUnion` around by value, why not change the `Point` implementations to instead use value receivers (since the methods don't mutate a `Point`), and just copy the Points directly.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/geo/geodist/geodist.go, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50790#-MB5totP2bpWBWm-Tc7x:-MB602kG6vPB_soUl8It:blde86g) ([raw file](https://github.com/cockroachdb/cockroach/blob/89a3dfa2466e9cc05923c9b2e68ce9631f2de6b2/pkg/geo/geodist/geodist.go#L32)):*
> It seems to me that the reason for the allocation is that the implementations of Point use a pointer receiver. If we are going to be copying PointUnion around by value, why not change the Point implementations to instead use value receivers (since the methods don't mutate a Point), and just copy the Points directly.

Not sure what you mean, does the method you suggest make the `Edge` functions for LinearRing / LineString not have to allocate a new pointer here?

Mind giving an example?


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/geo/geodist/geodist.go, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50790#-MB5totP2bpWBWm-Tc7x:-MB61EsmEFyZ9R_T1YKL:b-s2yoq4) ([raw file](https://github.com/cockroachdb/cockroach/blob/89a3dfa2466e9cc05923c9b2e68ce9631f2de6b2/pkg/geo/geodist/geodist.go#L32)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

> It seems to me that the reason for the allocation is that the implementations of Point use a pointer receiver. If we are going to be copying PointUnion around by value, why not change the Point implementations to instead use value receivers (since the methods don't mutate a Point), and just copy the Points directly.

Not sure what you mean, does the method you suggest make the `Edge` functions for LinearRing / LineString not have to allocate a new pointer here?

Mind giving an example?
</blockquote></details>

oh i see.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50790)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/geodist/geodist.go, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50790#-MB5totP2bpWBWm-Tc7x:-MBEN0Uj9iOlccpypoBn:b-6tmxg7) ([raw file](https://github.com/cockroachdb/cockroach/blob/89a3dfa2466e9cc05923c9b2e68ce9631f2de6b2/pkg/geo/geodist/geodist.go#L32)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

oh i see.
</blockquote></details>

Is this ready for another look?

I think my earlier suggestion to make the `Point` interface be implemented by value receivers for `s2GeodistPoint` and `geomDistPoint` was incorrect. I was imagining that the interface internal representation would be like a c++ union, but given the unknown number of structs that can implement an interface it has a pointer to the implementation of that interface. So it would have a pointer to `s2GeodistPoint` which means we can't really get "pass-by-value of the struct" behavior and there will still be the heap allocation issue. Is that consistent with your understanding?


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/geo/geodist/geodist.go, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50790#-MB5totP2bpWBWm-Tc7x:-MBF6ALrAO_bGd3_JGur:b-c6w9sq) ([raw file](https://github.com/cockroachdb/cockroach/blob/89a3dfa2466e9cc05923c9b2e68ce9631f2de6b2/pkg/geo/geodist/geodist.go#L32)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Is this ready for another look?

I think my earlier suggestion to make the `Point` interface be implemented by value receivers for `s2GeodistPoint` and `geomDistPoint` was incorrect. I was imagining that the interface internal representation would be like a c++ union, but given the unknown number of structs that can implement an interface it has a pointer to the implementation of that interface. So it would have a pointer to `s2GeodistPoint` which means we can't really get "pass-by-value of the struct" behavior and there will still be the heap allocation issue. Is that consistent with your understanding?

</blockquote></details>

Yeah it's ready.

Your evaluation is correct -- pass by value doesn't really work :(


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50790#-:-MBZT2U-Ee8DKyWj_JVq:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geodist/geodist.go, line 20 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50790#-MBZQWZ44dnxNiF3fRZH:-MBZQWZ5DtkAehTypN6e:b-wt387j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9df8a552ab9079a9f476aba0aee5aef8adba073/pkg/geo/geodist/geodist.go#L20)):*
> ```Go
> )
> 
> // Point is a union that avoids mallocing pointers for the heap.
> ```

... avoids mallocing objects from the heap.

How about something like the following
// Point is a union of the point types used in geometry and geography representation.
// The interfaces for distance calculation defined below are shared for both representations,
// and this union helps us avoid heap allocations by doing cheap copy-by-value of points. The
// code that peers inside a Point knows which of the two fields is populated.
___
*[pkg/geo/geodist/geodist.go, line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50790#-MBZSu9y1QuZH3R8odp5:-MBZSu9y1QuZH3R8odp6:b-y0wi83) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9df8a552ab9079a9f476aba0aee5aef8adba073/pkg/geo/geodist/geodist.go#L24)):*
> ```Go
> // Point interface structs whilst iterating over the shape.
> type Point struct {
> 	GeomCoord geom.Coord
> ```

How about s/GeomCoord/GeomPoint/ for naming consistency?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50790)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/geodist/geodist.go, line 20 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50790#-MBZQWZ44dnxNiF3fRZH:-MBZbjSY5rmj-DUqymOX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9df8a552ab9079a9f476aba0aee5aef8adba073/pkg/geo/geodist/geodist.go#L20)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

... avoids mallocing objects from the heap.

How about something like the following
// Point is a union of the point types used in geometry and geography representation.
// The interfaces for distance calculation defined below are shared for both representations,
// and this union helps us avoid heap allocations by doing cheap copy-by-value of points. The
// code that peers inside a Point knows which of the two fields is populated.
</blockquote></details>

Done.
___
*[pkg/geo/geodist/geodist.go, line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50790#-MBZSu9y1QuZH3R8odp5:-MBZbjjj84_cZCnK7HAG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9df8a552ab9079a9f476aba0aee5aef8adba073/pkg/geo/geodist/geodist.go#L24)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

How about s/GeomCoord/GeomPoint/ for naming consistency?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan committed:
geodist: use Point struct instead of a Point interface

Getting an Edge is expensive as the generating a Point that complies
with the geodist.Point interface involves mallocing from memory. We call
Edge(x) a lot when doing POLYGON <=> POLYGON comparisons.

Instead, change the Point to a struct, which does not require
any malloc to generate interfaces at the cost of code ugliness.

Without bounding box checks, this speeds up a query from 2min to
1min30sec on a ~120000 row join.

Release note: None

otan commented:
bors r=sumeerbhola
